### PR TITLE
Arc 1602 audience tech planning add four columns to the cashflow mart that represent previous year s and two year s ago data

### PR DESCRIPTION
**Pull request by Frydafly**
# Pull Request for dbt-arc-functions

- [x] Open the pull request as a draft so that you can run tests against the code and inspect the files in the Files view of the Pull Request interface in Github.

- [x] Once you're satisfied, convert the draft into an open Pull Request by clicking "Ready for Review" near the bottom of this page. You should not make any changes to the code except in dire emergency once you've converted from draft, so make sure you're really satisfied.

- [x] Add two of the principal contributors to this PR for review. Currently, the principal contributors are @dmbluestate , @Frydafly , and @ryantimjohn.

- [x] We've divided our PR templates into Macros, Utils and Docs PR requests. Delete the ones you're not using.
### Macros (delete if not using)

- [x] Link Jira tickets (or Github Issues) to PR here. Make sure that the Jira ticket has a good description as to why this PR is necessary and the business problem that this PR is solving. Moreover, if there have been any discussions of this ticket in Slack, make sure they're linked in the Jira ticket: https://bluestate.atlassian.net/browse/ARC-1602?atlOrigin=eyJpIjoiN2ViMDMwMzRiYmM5NDQ5YzhkMDFjMmZkMjI4MmM3NjgiLCJwIjoiaiJ9

- [x] Make sure this is named using Jira extension so that naming convention is consistent. If it's not, that's okay. Go to the Jira ticket, click into the ticket, then on the right under Details, find Development, then click the Create branch button. Copy the Branch Name from there. Read here [how to rename a branch in Github](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch) to rename this branch. In the future, create branches from Jira.

- [x] Make sure the Macro is documented by running the notebook: `create_docs.ipynb` util in this repo. From there, you'll have to either fill in descriptions by hand OR use our `fill_in_descriptions_using_openai.ipynb` to fill in descriptions using OpenAI's ChatGPT. Always make sure to check over ChatGPT's column descriptions, as they're likely not perfect.

- [x] Test branch on a development branch of an existing client (ideally the one that raised the bug). Do this by changing revision of the package to the branch name in `packages.yml` file in dbt State client below.

- [NA] If new macro files were created or updated, re-run `create_or_update_standard_models.py` for the client dbt project, replacing models and dependencies

- [x] check in dbt that client package successfully runs `dbt deps`, `dbt compile`, and `dbt run`; screenshot this page
Screenshot 2023-09-07 at 2 01 54 PM

- [x] query the resulting staging model in bigquery (or by previewing mart in dbt cloud, or your IDE); screenshot this

Screenshot 2023-09-07 at 2 35 09 PM

## Formatters

Three different formatters will run when you open a pull request:
- [sqlfmt](http://sqlfmt.com/)
- [yamlfix](https://lyz-code.github.io/yamlfix/)
- [Black(python)](https://pypi.org/project/black/)

If any of them detect syntax problems, they will error. They will automatically open a pull request with the changes you need to make to pass the formatter. Just merge that PR and you should be good to go.

## Linters
We have two in-house linters:

### [check_macros.py](https://github.com/bsd/dbt-arc-functions/blob/main/utils/check_macros.py)
Makes sure that every macro has a doc associated with it in the documentation folder with the same folder structure, i.e.:

`documentation/combined_email_paidmedia/marts/mart_combined_email_paidmedia_daily_revenue_performance.yml`

is the doc for

`       macros/combined_email_paidmedia/marts/mart_combined_email_paidmedia_daily_revenue_performance.sql`

Additionally makes sure that macros are named correctly, i.e. with the naming convention `create_(macro_file_name_stem)`

### [check_docs_correct.py](https://github.com/bsd/dbt-arc-functions/blob/main/utils/check_docs_correct.py)
Runs through a lot of checks on documents to make sure that they are formatted to our standards. This includes checking whether docs and sources:
- have columns
- have tables
- source columns have data_type and description
- have a version number
- docs have a macro associated with them
- docs are not blank
- that model names match the filename stem

There may be additional checks not documented here. The linter will fail on any error.

The script will document all failures and display them to you with remedies.

### CODE REVIEW

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget_1x_revenue.sql
Comment (dmbluestate) [inline macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget_1x_revenue.sql:1]: still seeing duplicates here.
Screenshot 2023-09-12 at 2 50 38 PM

**Review by dmbluestate — CHANGES_REQUESTED** [review]

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget_1x_revenue.sql
Comment (Frydafly) [reply macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget_1x_revenue.sql:1]: will work on that!

Comment (Frydafly) [reply macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget_1x_revenue.sql:1]: @dmbluestate addressed the duplicates

**Comment by Frydafly** [comment]
While addressing the duplicates, also addressed the missing information in cases where previous years data does not perfectly match up to the current transaction -- now there should be cases where we have a date, an audience, a channel, all without revenue or gifts or even a budget assigned to it, and if the previous year or two years from then, there was a value for revenue, it will show up as a mostly blank row with key information and will have the previous years revenue. 

@dmbluestate

File: macros/audience_budget/staging/stg_audience_budget_by_day_enriched.sql [hunk 1]
```diff
@@ -8,7 +8,7 @@
         date_spine_month,
         date_spine_day,
         {{ dbt_arc_functions.get_fiscal_year("date_day", var("fiscal_year_start")) }}
-        as fiscal_year,
+        as fiscal_year,  -- not sure if this is needed here yet or if we can just keep date_day
```
Comment (Frydafly) [inline macros/audience_budget/staging/stg_audience_budget_by_day_enriched.sql:11]: remove comment

File: macros/stitch_sfmc_arc/staging/stg_stitch_sfmc_arc_audience_union_transaction_joined_enriched.sql [hunk 2]
```diff
@@ -10,7 +10,7 @@
                 "audience_union_transaction_joined.transaction_date_day",
                 var("fiscal_year_start"),
             )
-        }} as fiscal_year,
+        }} as fiscal_year,  -- not sure if this field is needed just yet here
```
Comment (Frydafly) [inline macros/stitch_sfmc_arc/staging/stg_stitch_sfmc_arc_audience_union_transaction_joined_enriched.sql:13]: remove comment

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql [hunk 3]
```diff
@@ -7,52 +7,117 @@
...
+            select
+                donor_audience,
+                channel,
+                date_day,
+                total_revenue_actuals as prev_two_year_total_revenue_actuals,
+                total_revenue_budget_by_day as prev_two_year_total_revenue_budget
+            from dateoffset
+        ),
+        enriched as (
+            select
```
Comment (Frydafly) [inline macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql:77]: this date_day is correct for the dateoffset left join match, but the full outer join brings in the date_day one year too early or two years too early.

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql
Comment (dmbluestate) [inline macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql:1]: Screenshot 2023-09-13 at 3 28 34 PM

Still seeing some issues with the duplicate rows. The actuals are calculating correctly, but there's something wonky with the budget numbers and the date day.

**Review by dmbluestate — CHANGES_REQUESTED** [review]

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql
Comment (Frydafly) [reply macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql:1]: alrighty! reiterating here from our convo that the date_day value is the biggest issue, requiring an addition of one year for cases where we are looking at the prev one year value and an addition of two for the two year value.

Comment (Frydafly) [reply macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql:1]: @dmbluestate I made changes that I think might work, here is the equivalent query from above

```
SELECT * FROM `bsd-arc-uusa.dbt_fguedes_marts.mart_cashflow_actuals_and_budget` 
where month = 3
and day = 22
and donor_audience = 'grassroots'
and channel = 'email'
order by 5 desc
LIMIT 1000
```

would you be able to run this in bigquery and double check?

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql (hunk 3 above)
Comment (Frydafly) [reply macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql:77]: done!

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql [hunk 4]
```diff
@@ -88,14 +92,14 @@
                 coalesce(
                     dateoffset.channel, prevyear.channel, prevtwoyears.channel
                 ) as channel,
-                dateoffset.total_revenue_actuals,
+                dateoffset.total_revenue_actuals
```
Comment (dmbluestate) [inline macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql]: just a Missing comma here.

File: macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql
Comment (dmbluestate) [reply macros/stitch_sfmc_audience_transaction/marts/mart_cashflow_actuals_and_budget.sql:1]: Tested and can confirm all the math checks out. Still a remaining issue here for the 2025 year we still have two rows, one for the budget and one for the actuals. This only happens when filtering for a month and day that is before the fiscal year change over. So I tested for 3/22.4/22. and 6/15 and got the same results. One extra row for 2025 where the budget and actuals are on different rows, although they are calculating correctly.

**Review by dmbluestate — APPROVED** [review]

**Review by ryantimjohn — APPROVED** [review]

**Pull request merged**